File: HW1/hw1.java


/***
 *
 * @Author Kapil Sharma ks4643
 * Foundations of Artificial Intelligence
 * Homework 1-P
 * This program uses BFS to perform all the necessary action
 ***/

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;



public class hw1 {

    public static void main(String[] args) {
        if (args.length != 3) {
            System.out.println("Provide valid arguments");
            return;
        }

        String dictionaryPath = args[0];
        String firstWord = args[1];
        String secondWord = args[2];

        try {
            Set<String> words = readDictionary(dictionaryPath);
            List<String> wordList = filterWords(words, firstWord.length());
            Map<String, List<String>> graph = simulateGraph(wordList);

            List<String> shortestPath = bfs(graph, firstWord, secondWord);

            if (shortestPath != null) {
                for (String word : shortestPath) {
                    System.out.println(word);
                }
            } else {
                System.out.println("No solution");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static Set<String> readDictionary(String filePath) throws IOException {
        Set<String> words = new HashSet<>();
        BufferedReader reader = new BufferedReader(new FileReader(filePath));
        String line;
        while ((line = reader.readLine()) != null) {
            words.add(line.trim().toLowerCase());
        }
        reader.close();
        return words;
    }

    private static List<String> filterWords(Set<String> words, int length) {
        List<String> filteredWords = new ArrayList<>();
        for (String word : words) {
            if (word.length() == length) {
                filteredWords.add(word);
            }
        }
        return filteredWords;
    }

    private static Map<String, List<String>> simulateGraph(List<String> wordList) {
        Map<String, List<String>> graph = new HashMap<>();
        for (String word : wordList) {
            graph.put(word, new ArrayList<>());
            for (String nextWord : wordList) {
                if (letterComparisons(word, nextWord)) {
                    graph.get(word).add(nextWord);
                }
            }
        }
        return graph;
    }

    private static boolean letterComparisons(String word1, String word2) {
        int difference = 0;
        for (int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) {
                difference++;
            }
        }
        return difference == 1;
    }

    private static List<String> bfs(Map<String, List<String>> graph, String start, String end) {
        Queue<String> queue = new LinkedList<>();
        queue.add(start);
        Map<String, String> predecessor = new HashMap<>();
        predecessor.put(start, null);

        while (!queue.isEmpty()) {
            String current = queue.poll();
            if (current.equals(end)) {
                break;
            }

            for (String neighbor : graph.get(current)) {
                if (!predecessor.containsKey(neighbor)) {
                    predecessor.put(neighbor, current);
                    queue.add(neighbor);
                }
            }
        }

        if (predecessor.containsKey(end)) {
            List<String> path = new ArrayList<>();
            String current = end;
            while (current != null) {
                path.add(0, current);
                current = predecessor.get(current);
            }
            return path;
        } else {
            return null;
        }
    }
}

File: HW1/hw1.py
"""
Foundations Of Artificial Intelligence
Homework 1-P
Author: @Yash Mahajan (ym9800)
"""
import sys
from queue import Queue


def difference_in_letters(word1, word2):
    difference = 0
    for count1, count2 in zip(word1, word2):
        if count1 != count2:
            difference += 1
    return difference == True

def create_graph(word_list):
    graph = {}
    for i in range(len(word_list)):
        graph[word_list[i]] = []
        for next_word in word_list:
            if difference_in_letters(word_list[i], next_word):
                graph[word_list[i]].append(next_word)
    return graph

def bfs(graph, start, end):
    queue = Queue()
    queue.put(start)
    predecessor = {}
    predecessor[start] = None

    while not queue.empty():
        current = queue.get()
        if current == end:
            break

        for neighbor in graph[current]:
            if neighbor not in predecessor:
                predecessor[neighbor] = current
                queue.put(neighbor)

    if end in predecessor:
        path = []
        current = end
        while current != start:
            path.insert(0, current)
            current = predecessor[current]
        path.insert(0, start)
        return path
    else:
        return None

def main(dictionary_path, first_word, second_word):
    words = set()
    with open(dictionary_path, 'r') as file:
        for word in file:
            words.add(word.strip().lower())
    list_of_words = [word for word in words if len(word) == len(first_word)]
    graph = create_graph(list_of_words)
    shortest_path = bfs(graph, first_word, second_word)
    if shortest_path:
        for word in shortest_path:
            print(word)
    else:
        print("No solution")

if __name__ == "__main__":
        main(sys.argv[1], sys.argv[2], sys.argv[3])